README.md

# ParaBox.js

Toolbox for parallax web page.

## Features

- Background Image Parallax
- Item Position Parallax
- Define sections and trigger "sectionChange" event

## Usage

Basically, pass action's name and extra arguments to $.fn.parabox()

```javascript
$(/* some element */).parabox( action, options );
```

### Background

Update background image's top position for scrolling.

```javascript
$(".element-which-has-background").parabox("background", {
	bgHeight: 480,
	area: 240,
	reverse: true 
});
```

- bgHeight:Integer (480) : Background image height
- area:Integer (240) : Area which background animate (distance to offset top of its node)
- reverse:Boolean (true) : Animate reversely or not

### Item

Animate item element when scroll top enters or leaves the area

```javascript
$(".element-to-animte-for-scrolling").parabox("item", {
	top: 0,
	bottom: $("body").attr("offsetHeight"),
	easing: "swing",
	duration: 500,
	from: { left: -100, opactiy: 0 },
	to: { left: 0, opacity: 1 },
    done: function(show){
        console.log(show ? "show" : "hide");
    }
}
});
```

- top:Integer (null) : Scroll top align which activates item
- bottom:Integer (null) : Scroll bottom align which deactivates item
- easing:String ("swing") : Easing function name
- duration:Integer (500) : Duration time for animation
- delay:Integer (0) : Delay for starting animation
- from:Object ({}) : Styles when it's deactive
- to:Object ({}) : Styles when it's active
- done:Function (function(){}) : Handler for completing animation

When `bottom` is null, it will be set page's height

### Section

Define the named section and vertical align, 
and add event which trigger when section changes.

```javascript
$(document).parabox("section", {
    "top": 0,
    "section-first": 500,
    "section-second" 1000,
    "section-third": 1500,
    ...
})
```

The action "section" initialize the sections in page by passed object,
which consists of name as key and vertical align as value.

```javascript
$(document).on("sectionChange", function(e, data){
    console.log(data.index); // index for the section
    console.log(data.name); // key for the section
    console.log(data.value); // vertical align for the section
    console.log(data.instance); // $.ParaBox.section object
});
```

Event handler for "sectionChange" accept data which contains some values about section,
as second argument.



doc/changelog.md

# Change Log


## 0.9.2 (2013-09-02)

- Add `done` handler for completeing item animation

## 0.9.1 (2013-08-16)

- Add option "delay" for item

src/parabox.js
(function($){

	/**
	 * ParaBox
	 * -------
	 * Collection of tools for parallax page
	 */
	$.ParaBox = {};

	/**
	 * Background parallax
	 * -------------------
	 */
	$.ParaBox.background = {

		/**
		 * Defaults for options:
		 * - bgHeight:Integer = Background images's height
		 * - area:Integer = Area which background moves, margin from offset top of element
		 * - reverse:Boolean = Move background reversely or not
		 */
		defaults: {
			bgHeight: 480,
			area: 240,
			reverse: true
		},

		/**
		 * Initialize background parallax
		 * @param jQueryObject nodes
		 * @param Object options
		 */
		initialize: function(nodes, options){
			var my = this;
			options = $.extend({}, this.defaults, options);
			nodes.each(function(){
				var node, attrs;
				node = $(this);
				attrs = {
					reverse : options.reverse,
					height : node.prop("clientHeight"),
					left : $.ParaBox.util.getLeft(node),
					offset : node.offset().top,
					area : options.area
				};
				attrs.margin = options.bgHeight - attrs.height;
				node.data("paraBoxAttrs", attrs);
				$(window).on("scroll", $.proxy(my.onScroll, node));
			});
		},

		/**
		 * Hanlder for window scrolls
		 * `this` is assigned as the each node (jQueryObject)
		 * Background moves its position for scrolling
		 */
		onScroll: function(){
			var attrs, diff, top;
			attrs = this.data("paraBoxAttrs");
			diff = $(window).scrollTop() - attrs.offset;
			if(Math.abs(diff) < attrs.area){
				top = attrs.reverse ? ((diff + attrs.area) / (attrs.area * 2) * attrs.margin * -1) + "px"
				: ((1 - (diff + attrs.area) / (attrs.area * 2)) * attrs.margin * -1) + "px";
			} else if(attrs.reverse) {
				top = diff < 0 ? "top" : "bottom";
			} else {
				top = diff < 0 ? "bottom": "top";
			}
			this.css("background-position", [attrs.left, top].join(" ") );
		}

	};

	/**
	 * Item parallax
	 * -------------
	 */
	$.ParaBox.item = {

		/**
		 * Defaults for options:
		 * - top:Integer = Top align for the area which activates the item
		 * - bottom:Integer = Bottom align for the area which activates the item
		 * - easing:String = Easing function's name for animation
		 * - duration:Integer = Duration time for animation
		 * - delay:Integer = Delay time for starting animation
		 * - from:Object = CSS when deactivated
		 * - to:Object = CSS when activated
		 * - done:Function = Handler for animation completeing `function(show, promise, jumbed)`
		 */
		defaults: {
			top: null,
			bottom: null,
			easing: "swing",
			duration: 500,
			delay: 0,
			from: {},
			to: {},
			done: function(){}
		},

		/**
		 * Initialize item parallax
		 * @param jQueryObject nodes
		 * @param Object options
		 */
		initialize: function(nodes, options){
			var my = this;
			options = $.extend(true, {}, this.defaults, options);
			if(! options.bottom){
				options.bottom = $("body").height();
			}
			nodes.each(function(){
				var node = $(this);
				node.data("paraBoxAttrs", options);
				node.data("paraBoxShow", false);
				node.css(options.from);
				$(window).on("scroll", $.proxy(my.onScroll, node));
			});
		},

		/**
		 * Handler for window scrolls
		 * `this` is assigned as the each node (jQueryObject)
		 * The node animates to from/to styles when the activated status changes
		 */
		onScroll: function(){
			var scroll, attrs, show, inRange, done;
			scroll = $(window).scrollTop();
			attrs = this.data("paraBoxAttrs");
			show = this.data("paraBoxShow");
			done = function(){
				var node, args;
				node = $(this);
				args = [node.data("paraBoxShow")].concat([].slice.call(arguments));
				node.data("paraBoxAttrs").done.apply(this, args);
			};
			inRange = $.ParaBox.util.inRange(scroll, attrs.top, attrs.bottom);
			if((show && ! inRange) || (! show && inRange)){
				this.stop().delay(attrs.delay).animate(show ? attrs.from : attrs.to, {
					easing: attrs.easing,
					duration: attrs.duration,
					done: done
				});
				this.data("paraBoxShow", ! show);
			}
		}
	};

	/**
	 * sectionChange Event
	 * -------------------
	 */
	$.ParaBox.section = {

		/**
		 * Attributes:
		 * - node:jQueryObject = The node which trigger the event
		 * - points:Array = Collection of section name and its scroll top
		 * - index:Integer = Section index which is activated now
		 */
		node: null,
		points: null,
		index: null,

		/**
		 * Initialize sections and add handler
		 * `options` is object which contains labeled align, like `{name: align}`
		 * @param jQueryObject node
		 * @param Object options
		 */
		initialize: function(node, options){
			var points = [];
			$.each(options, function(name, value){
				points.push({
					name: name,
					value: value
				});
			});
			points.sort(function(a, b){
				return a.value - b.value;
			});
			this.points = points;
			this.node = node;
			$(window).on("scroll", $.proxy(this.onScroll, this));
		},

		/**
		 * Hanlder for window scrolls
		 * `this` is assigned as $.ParaBox.section
		 * When scroll top leap over some section's, 
		 * this trigger "sectionChange" event on `this.node`
		 */
		onScroll: function(){
			var i, scroll, point, data;
			scroll = $(window).scrollTop();
			i = this.points.length;
			while(i--){
				if(this.points[i].value < scroll){ break; }
			}
			if(i !== this.index){
				point = this.points[i];
				data = {
					index: i,
					name: point ? point.name : null,
					value: point ? point.value : null ,
					instance: this
				};
				this.node.trigger("sectionChange", data, this);
				this.index = i;
			}
		}
	};

	/**
	 * Utilities 
	 * ---------
	 */
	$.ParaBox.util = {

		/**
		 * Returns whether `num` is in the range `min` to `max`
		 * @param Number num
		 * @param Number min
		 * @param Number max
		 */
		inRange: function(num, min, max){
			return num >= min && num < max;
		},

		/**
		 * Helper to get left position of background image
		 * @param jQueryObject node
		 */
		getLeft: function(node){
			return node.css("background-position-x")
			|| node.css("background-position").split(" ")[0];
		}

	};

	/**
	 * Interface as jQuery feature
	 */
	$.fn.parabox = function(/* action, arg1, arg2 ... */){
		var args, action, obj;
		args = $.makeArray(arguments);
		action = args.shift();
		obj = $.ParaBox[action];
		if(obj){
			args.unshift(this);
			obj.initialize.apply(obj, args);
		}
		return this;
	};

}(jQuery));
